File: matlab/src/lang/constant/PollingType.m
classdef (Sealed) PollingType
    % Enumeration of polling service types
    %
    % Copyright (c) 2012-2026, Imperial College London
    % All rights reserved.
    
    properties (Constant)
        GATED = 0;
        EXHAUSTIVE = 1;
        KLIMITED = 2;
        DECREMENTING = 3;
    end
    
    methods (Static)        
        function text = toText(type)
            % TEXT = TOTEXT(TYPE)
            switch type
                case PollingType.GATED
                    text = 'Gated';
                case PollingType.EXHAUSTIVE
                    text = 'Exhaustive';
                case PollingType.KLIMITED
                    text = 'K-Limited';
                case PollingType.DECREMENTING
                    text = 'Decrementing';
            end
        end

        function id = toId(type)
            % ID = TOID(TYPE)
            id = type;
        end

        function name = toName(type)
            % NAME = TONAME(TYPE) - Canonical name used on the JSON wire.
            % The numeric ids agree with the Java enum but not with the Python
            % one, which assigns them via auto(), so names are used to
            % interchange polling types across codebases.
            switch PollingType.toId(type)
                case PollingType.GATED
                    name = 'GATED';
                case PollingType.EXHAUSTIVE
                    name = 'EXHAUSTIVE';
                case PollingType.KLIMITED
                    name = 'KLIMITED';
                case PollingType.DECREMENTING
                    name = 'DECREMENTING';
                otherwise
                    line_error(mfilename, sprintf('Unable to return a PollingType name for value: %d.\n', PollingType.toId(type)));
            end
        end

        function id = fromName(name)
            % ID = FROMNAME(NAME) - Inverse of toName.
            switch upper(name)
                case 'GATED'
                    id = PollingType.GATED;
                case 'EXHAUSTIVE'
                    id = PollingType.EXHAUSTIVE;
                case 'KLIMITED'
                    id = PollingType.KLIMITED;
                case 'DECREMENTING'
                    id = PollingType.DECREMENTING;
                otherwise
                    line_error(mfilename, sprintf('Unable to return a PollingType for name: %s.\n', name));
            end
        end
    end
end

File: jar/src/main/kotlin/jline/lang/constant/PollingType.java
/*
 * Copyright (c) 2012-2026, QORE Lab, Imperial College London
 * All rights reserved.
 */

package jline.lang.constant;

/**
 * Enumeration of polling service types for polling scheduling strategy.
 * <p>
 * These constants specify how a polling server handles jobs at each queue
 * during a polling cycle:
 * <ul>
 * <li>GATED - Serves only jobs present at the start of the polling cycle</li>
 * <li>EXHAUSTIVE - Serves all jobs until the queue is empty</li>
 * <li>KLIMITED - Serves at most K jobs per polling cycle</li>
 * </ul>
 */
public enum PollingType {
    GATED(0),
    EXHAUSTIVE(1),
    KLIMITED(2);

    private final int value;

    PollingType(int value) {
        this.value = value;
    }

    /**
     * Gets the numeric value of this polling type.
     *
     * @return the numeric value
     */
    public int getValue() {
        return value;
    }

    /**
     * Converts a numeric value to a PollingType enum value.
     *
     * @param value the numeric value
     * @return the corresponding PollingType enum value
     * @throws RuntimeException if the value is not recognized
     */
    public static PollingType fromValue(int value) {
        switch (value) {
            case 0:
                return GATED;
            case 1:
                return EXHAUSTIVE;
            case 2:
                return KLIMITED;
            default:
                throw new RuntimeException("Unable to return a PollingType for value: " + value);
        }
    }

    /**
     * Converts a PollingType enum value to its text representation.
     *
     * @param type the polling type to convert
     * @return the text representation of the type
     */
    public static String toText(PollingType type) {
        switch (type) {
            case GATED:
                return "Gated";
            case EXHAUSTIVE:
                return "Exhaustive";
            case KLIMITED:
                return "K-Limited";
            default:
                return "";
        }
    }

    /**
     * Gets the ID (numeric value) of the polling type.
     *
     * @param type the polling type
     * @return the ID value
     */
    public static int toId(PollingType type) {
        return type.getValue();
    }
}